torch/csrc/jit/script/final_returns.cpp

```diff
@@ -1,106 +1,270 @@
+    while (iter != block->nodes().end()) {
+      auto node = *iter++;
+      node->moveBefore(guard_block->return_node());
+    }
+
+    // Set the new if to have the same outputs of the original block,
+    // then replace the original block outputs with new if's outputs
+    for (size_t i = 0; i < block->outputs().size(); ++i) {
+      return_block->registerOutput(bottom_val);
+      guard_block->registerOutput(block->outputs().at(i));
+      new_if->addOutput()
+          ->setType(block->outputs().at(i)->type())
+          ->setUniqueName(block->outputs().at(i)->uniqueNameBase());
+    }
+    while (block->outputs().size() > 0) {
+      block->eraseOutput(0);
+    }
+    for (auto out : new_if->outputs()) {
+      block->registerOutput(out);
+    }
+
+    return handleIf(new_if);
+  }
+
+  void deleteAfterReturnNodes(Block* block, graph_node_list_iterator& iter) {
+    if (iter == block->nodes().end()) {
+      return;
+    }
+    // need to destroy in reverse order so nodes have no uses when destroyed
+    for (auto it = block->nodes().reverse().begin(); it != iter;) {
+      if (*it == block->return_node()) {
+        it++;
+      } else {
+        it.destroyCurrent();
+      }
+    }
+    iter->destroy();
+  }
+
+  void checkNoLoopReturn(Block* block) {
+    for (Node* n : block->nodes()) {
+      if (n->kind() == prim::ReturnStmt) {
+        throw ErrorReport(n->getSourceLocation())
+            << "Return statements within loops are not yet supported\n";
+      }
+      for (Block* b : n->blocks()) {
+        checkNoLoopReturn(b);
+      }
+    }
+  }
+
+  ReturnStatus makeReturnsFinal(Block* block) {
+    auto ret_status = WONT_RETURN;
+    for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+      Node* node = *it;
+      it++;
+      switch (node->kind()) {
+        case prim::ReturnStmt: {
+          ret_status = WILL_RETURN;
+          block_return_vals_[block] = node->input();
+          node->destroy();
+        } break;
+        case prim::If: {
+          ret_status = handleIf(node);
+        } break;
+        case prim::Loop:
+          checkNoLoopReturn(node->blocks().at(0)); // not yet supported
+        default:
+          break;
+      }
+      if (ret_status == WILL_RETURN) {
+        deleteAfterReturnNodes(block, it);
         break;
-      case TK_RETURN:
-        changed.emplace_back(stmt);
-        // ignore the rest the the block, it is dead.
-        return {true, List<Stmt>::unsafeCreate(range, std::move(changed))};
-      default:
-        changed.emplace_back(stmt);
+      } else if (ret_status == MIGHT_RETURN) {
+        if (it != block->nodes().end()) {
+          ret_status = guardBlockNodes(
+              block, getHasReturned(node), getReturnVal(node), it);
+        }
         break;
+      }
+    }
+    if (block_status_.count(block) == 0) {
+      block_status_[block] = ret_status;
+      return ret_status;
+    } else {
+      // Guarded return blocks have their status set prior
```

Comment:
Answer 'why' not 'what'. I can see that return status is already set, but I don't know why.